File: Baitap2/Complete/Intergration Testing/code/MergeSort.py
def __validate_(arr):
    if (type(arr) is not list):
        return False
    for x in arr:
        if (type(x) is not int):
            return False
    return True

def merge(arrLeft, arrRight):
    if (not __validate_(arrLeft)):
        raise ValueError("arrLeft is not a list of int!")
    if (not __validate_(arrRight)):
        raise ValueError("arrRight is not a list of int!")
    lenL = len(arrLeft); lenR = len(arrRight)
    result = []; idxL = 0; idxR = 0
    
    while idxL < lenL and idxR < lenR:
        if (arrLeft[idxL] < arrRight[idxR]):
            result.append(arrLeft[idxL])
            idxL += 1
        else:
            result.append(arrRight[idxR])
            idxR += 1
    while idxL < lenL:
        result.append(arrLeft[idxL])
        idxL += 1
    while idxR < lenR:
        result.append(arrRight[idxR])
        idxR += 1

    return result

def sort(arr):
    if (not __validate_(arr)):
        raise ValueError("arr is not a list of int!")
    if (len(arr) <= 1):
        return arr
    middle = len(arr) // 2
    arrLeft = sort(arr[0: middle])
    arrRight = sort(arr[middle: len(arr)])
    result = merge(arrLeft, arrRight)
    return result

File: Baitap2/Complete/Intergration Testing/code/MergeSort.java
import java.util.ArrayList;
import java.util.Scanner;

class MergeSort{
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        
        System.out.println("input a list:");
        int[] arr = new int[100];
        int n = scan.nextInt();
        for(int i = 0 ; i < n ; i++){
            arr[i] = scan.nextInt();
        }

        if(n==0) return;
        
        merge_sort(arr, 0, n-1);
        
        System.out.println("sorted list:");
        for(int i = 0 ; i < n ; i++){
            System.out.print(arr[i] + " ");
        }
    }

    public static void merge_sort(int[] arr , int left , int right){
        if(left == right) return;
        
        int mid = (right + left)/2;
        merge_sort(arr, left, mid);
        merge_sort(arr, mid + 1, right);
        
        merge(arr , left , mid , right);
        return;
    }

    public static void merge(int[] arr, int left, int mid, int right){
        int idxL = left; 
        int idxR = mid + 1;
        
        int[] newarr = new int[100];
        int idx = 0;

        while(idxL <= mid && idxR <= right){
            if(arr[idxL] < arr[idxR]){
                newarr[idx++] = arr[idxL++];
            }
            else{
                newarr[idx++] = arr[idxR++];
            }
        }

        while(idxL <= mid){
            newarr[idx++] = arr[idxL++];
        }
        
        while(idxR <= right){
            newarr[idx++] = arr[idxR++];
        }

        idx=0;
        while(left<=right){
            arr[left++] = newarr[idx++];
        }

        return;
    }
}